File: java/src/main/java/com/example/cote/PGMRS/운영체제.java
package com.example.cote.PGMRS;

import java.util.*;
import java.io.*;
import java.nio.file.*;
import java.util.stream.Collectors;

public class 운영체제 {
	private static final int MAX_PROGRAMS_LENGTH = 100000;
	private static final int MAX_PRIORITY = 10;
	private static final int MAX_TIMES = 10000000;
	private static final int MAX_RUNTIMES = 1000;

	private static class Result{

		public int completedTime;
		public int[] totalWaitedTimes = new int[10];

		Result(){}

		Result(int completedTime, int[] totalWaitedTimes) {
			this.completedTime = completedTime;
			this.totalWaitedTimes = totalWaitedTimes;
		}

		@Override 
		public String toString() {
			return String.format("(%d) %s", completedTime, Arrays.toString(totalWaitedTimes));
		}
	}
	private static record Program( 
		int priority,
		int calledAt,
		int executionTime
	){}


	private static Result solve(int[][] programs) {
		Queue<Program> queue = Arrays.stream(programs)
			.map(p -> new Program(p[0], p[1], p[2]))
			.sorted(Comparator.comparingInt(Program::calledAt))
			.collect(Collectors.toCollection(LinkedList::new));

		PriorityQueue<Program> pq = new PriorityQueue<>((a, b) -> {
			if (a.priority() != b.priority()) {
				return a.priority() - b.priority();
			}
			return a.calledAt() - b.calledAt();
		});

		Result ret = new Result();
		int time = 0;

		while (!queue.isEmpty() || !pq.isEmpty()) {
			while (!queue.isEmpty() && time >= queue.peek().calledAt()) {
				pq.add(queue.poll());
			}

			if (pq.isEmpty()) {
				time = queue.peek().calledAt();
				continue;
			}

			Program p = pq.poll();
			ret.totalWaitedTimes[p.priority()] += time - p.calledAt();
			time += p.executionTime();
		}
		ret.completedTime = time;
			
		return ret;
	}

	public static void main(String... args) throws IOException {
		Path p = Paths.get(System.getProperty("user.dir") + "/data/운영체제.txt");
		BufferedReader reader = Files.newBufferedReader(p);

		int testCases = Integer.parseInt(reader.readLine());

		while (testCases > 0) {
			int numPrograms = Integer.parseInt(reader.readLine());


			int[][] programs = new int[numPrograms][3];

			int[] priorities = Arrays.stream(reader.readLine().split(" "))
					.mapToInt(Integer::parseInt)
					.toArray();

			int[] times = Arrays.stream(reader.readLine().split(" "))
					.mapToInt(Integer::parseInt)
					.toArray();

			int[] runtimes = Arrays.stream(reader.readLine().split(" "))
					.mapToInt(Integer::parseInt)
					.toArray();

			for (int i = 0; i < numPrograms; i++) {
				programs[i][0] = priorities[i] - 1; 
				programs[i][1] = times[i]; 
				programs[i][2] = runtimes[i]; 
			}

			Result answer = new Result();

			answer.completedTime = Integer.parseInt(reader.readLine());
			answer.totalWaitedTimes = Arrays.stream(reader.readLine().split(" "))
				.mapToInt(Integer::parseInt)
				.toArray();

			Result solved = solve(programs);

			System.out.println("answer : " + answer);
			System.out.println("solved : " + solved);

			testCases--;
		}
	}
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/운영체제.kt
package com.example.cote.PGMRS

import java.io.*
import java.nio.file.*
import java.util.*

object 운영체제 {

    val MAX_PROGRAMS_LENGTH = 100000
    val MAX_PRIORIRY = 10
    val MAX_TIMES = 1000000
    val MAX_RUNTIMES = 1000

    data class Answer(
        val completedTime: Int = 0,
        val totalWaitedTimes: Array<Int> = Array(MAX_PRIORIRY) { 0 }
    )

    data class Program(val priority: Int, val time: Int, val runtime: Int)

    fun solve(programs: List<Array<Int>>): Answer {
        val sortedPrograms =
            programs
                .map { Program(it[0] - 1, it[1], it[2]) }
                .sortedBy(Program::time)

        val q = ArrayDeque<Program>(sortedPrograms)

        val pq =
            PriorityQueue<Program>(compareBy(Program::priority, Program::time))

        var current = 0
        val waitedTimes = Array<Int>(MAX_PRIORIRY) { 0 }

        while (!q.isEmpty() || !pq.isEmpty()) {

            while (!q.isEmpty() && q.first().time <= current) {
                pq.offer(q.removeFirst())
            }

            if (pq.isEmpty()) {
                current = q.peek().time
                continue
            }

            val (priority, time, runtime) = pq.poll()
            waitedTimes[priority] += current - time
            current += runtime
        }

        return Answer(current, waitedTimes)
    }

    fun main() {
        val reader =
            with(Paths.get(System.getProperty("user.dir") + "/data/운영체제.txt")) {
                Files.newBufferedReader(this)
            }

        repeat(reader.readLine().toInt()) {
            val numPrograms = reader.readLine().toInt()
            val programs = MutableList(numPrograms) { Array<Int>(3) { 0 } }
            val priorities = reader.readLine().split(" ").map(String::toInt)
            val times = reader.readLine().split(" ").map(String::toInt)
            val runtimes = reader.readLine().split(" ").map(String::toInt)

            for (i in 0..numPrograms - 1) {
                programs[i] = arrayOf(priorities[i], times[i], runtimes[i])
            }

            val completedTime = reader.readLine().toInt()
            val totalWaitedTimes =
                reader.readLine().split(" ").map(String::toInt).toTypedArray()
            val answer = Answer(completedTime, totalWaitedTimes)
            println(answer)
            println(solve(programs))
            println()
        }
    }
}

fun main() {
    운영체제.main()
}
